crl/updater/updater.go

```diff
@@ -0,0 +1,364 @@
+
+	capb "github.com/letsencrypt/boulder/ca/proto"
+	"github.com/letsencrypt/boulder/issuance"
+	blog "github.com/letsencrypt/boulder/log"
+	sapb "github.com/letsencrypt/boulder/sa/proto"
+)
+
+type crlUpdater struct {
+	issuers           map[issuance.IssuerNameID]*issuance.Certificate
+	numShards         int
+	lookbackPeriod    time.Duration
+	lookforwardPeriod time.Duration
+	updatePeriod      time.Duration
+	maxParallelism    int
+
+	sa sapb.StorageAuthorityClient
+	ca capb.CRLGeneratorClient
+	// TODO(#6162): Add this so we can talk to the crl-storer.
+	// cs cspb.CRLStorerClient
+
+	tickHistogram    *prometheus.HistogramVec
+	generatedCounter *prometheus.CounterVec
+
+	log blog.Logger
+	clk clock.Clock
+}
+
+func NewUpdater(
+	issuers []*issuance.Certificate,
+	numShards int,
+	lookbackPeriod time.Duration,
+	lookforwardPeriod time.Duration,
+	updatePeriod time.Duration,
+	maxParallelism int,
+	sa sapb.StorageAuthorityClient,
+	ca capb.CRLGeneratorClient,
+	stats prometheus.Registerer,
+	log blog.Logger,
+	clk clock.Clock,
+) (*crlUpdater, error) {
+	issuersByNameID := make(map[issuance.IssuerNameID]*issuance.Certificate, len(issuers))
+	for _, issuer := range issuers {
+		issuersByNameID[issuer.NameID()] = issuer
+	}
+
+	if numShards < 1 {
+		return nil, fmt.Errorf("must have positive number of shards, got: %d", numShards)
+	}
+
+	if updatePeriod >= 7*24*time.Hour {
+		return nil, fmt.Errorf("must update CRLs at least every 7 days, got: %s", updatePeriod)
+	}
+
+	window := lookbackPeriod + lookforwardPeriod
+	if window.Nanoseconds()%int64(numShards) != 0 {
+		return nil, fmt.Errorf("total window (lookback+lookforward=%dns) must be evenly divisible by numShards (%d)",
+			window.Nanoseconds(), numShards)
+	}
+
+	if maxParallelism <= 0 {
+		maxParallelism = 1
+	}
+
+	tickHistogram := prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "crl_updater_ticks",
+		Help:    "A histogram of crl-updater tick latencies labeled by issuer and result",
+		Buckets: []float64{0.01, 0.2, 0.5, 1, 2, 5, 10, 20, 50, 100, 200, 500, 1000, 2000, 5000},
+	}, []string{"issuer", "result"})
+	stats.MustRegister(tickHistogram)
+
+	generatedCounter := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "crl_updater_generated",
+		Help: "A counter of CRL generation calls labeled by result",
+	}, []string{"result"})
+	stats.MustRegister(generatedCounter)
+
+	// TODO(#6162): add a storedCounter when sending to the crl-storer.
+
+	return &crlUpdater{
+		issuersByNameID,
+		numShards,
+		lookbackPeriod,
+		lookforwardPeriod,
+		updatePeriod,
+		maxParallelism,
+		sa,
+		ca,
+		tickHistogram,
+		generatedCounter,
+		log,
+		clk,
+	}, nil
+}
+
+// Run causes the crl-updater to run immediately, and then re-run continuously
+// on the frequency specified by crlUpdater.updatePeriod.
+func (cu *crlUpdater) Run() {
+	cu.tick(context.Background())
+	ticker := time.Tick(cu.updatePeriod)
+	for range ticker {
+		cu.tick(context.Background())
```

Comment:
Same comment as above.

```suggestion
		cu.tick(context.Background())
```